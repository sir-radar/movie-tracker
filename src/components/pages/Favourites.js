import React,{useEffect} from 'react';
import { connect } from 'react-redux';
import MovieCard from '../MovieCard';
import Loader from '../Loader';
import NoData from '../NoData';
import { getAllFavourites, removeFromFavourite } from '../../store/actions/favouritesActions';

const Favourites = ({favouriteMovies, removeFavourite, getAllFavourites, favouriteIDs, favouriteStatus, watchlistIDs}) => {
  //request for all favourites
  useEffect(() => {
    getAllFavourites()
  },[getAllFavourites]);

  return (
    <div className="row px-5">
      {
        //display on successful request
        (favouriteStatus === 'SUCCESS') ? favouriteMovies.results.map(result => <MovieCard
            key={result.id}
            id={result.id}
            title={result.title}
            image={result.poster_path}
            overview={result.overview}
            removeFavourite={removeFavourite}
            favouriteIDs={favouriteIDs}
            watchlistIDs={watchlistIDs}
          />) 
        
        : null
      }

      {
        //display while request is pending
        (favouriteStatus === 'PENDING')
        ? <Loader/>
        : null
      }
      
      {
        //display if there is no favourite
        (favouriteStatus === 'SUCCESS' && favouriteMovies.results.length < 1)
        ? <NoData/>
        : null
      }

  
    
    </div>
  );
}

const mapStateToProps = (state) => (
  {
    favouriteStatus: state.status.favourite,
    favouriteError: state.status.favouriteError,
    favouriteMovies: state.favourites.favouriteMovies,
    favouriteIDs: state.favourites.favouriteMoviesID,
    watchlistIDs: state.watchlists.watchListsID,
  }
);
  
const mapDispatchToProps = (dispatch) => (
  {
    getAllFavourites: () => dispatch(getAllFavourites()),
    removeFavourite: (payload) => dispatch(removeFromFavourite(payload))
  }
)

export default connect(mapStateToProps, mapDispatchToProps) (Favourites);
