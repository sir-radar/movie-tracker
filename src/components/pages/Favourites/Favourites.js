import React,{useEffect} from 'react';
import { connect } from 'react-redux';
import PageLayout from '../../layout/PageLayout';
import MovieCard from '../../MovieCard/MovieCard';
import Loader from '../../Loader/Loader';
import NoData from '../../NoData/NoData';
import Error from '../../Error/Error';
import ErrorMessage from '../../ErrorMessage/ErrorMessage';
import Pagination from '../../Pagination/Pagination';
import { getAllFavourites, removeFromFavourite } from '../../../store/actions/favouritesActions';
import { addToWatchList, removeFromWatchList } from '../../../store/actions/watchListActions';
import { resetStatus } from '../../../store/actions/statusActions';

const Favourites = (props) => {
  const {
          favouriteMovies, 
          removeFavourite, 
          getAllFavourites, 
          favouriteStatus, 
          removeFromWatchList,
          addToWatchList,
          watchlistActionStatus,
          favouriteActionStatus,
          resetStatus
        } = props;

  //request for all favourites
  useEffect(() => {
    getAllFavourites()
  },[getAllFavourites]);

  
  return (
    <PageLayout

      showNav = {true}

      pageContent = {
        //display on successful request
        (favouriteStatus === 'SUCCESS') ? favouriteMovies.results.map(result => <MovieCard
            key={result.id}
            id={result.id}
            title={result.title}
            image={result.poster_path}
            overview={result.overview}
            removeFavourite={removeFavourite}
            removeFromWatchList={removeFromWatchList}
            addToWatchList={addToWatchList}
          />) 
        
        : null
      }

      loader = {
        //display while request is pending
        (favouriteStatus === 'PENDING')
        ? <Loader/>
        : null
      }

      error = {
        //display when an error occurs
        (favouriteStatus === 'ERROR')
        ? <Error/>
        : null
      }

      errorMessage = {
        //display when an error occurs in favouriting or adding movie to watchlist
        (watchlistActionStatus === 'ERROR' || favouriteActionStatus === 'ERROR' )
        ? <ErrorMessage resetStatus={resetStatus}/>
        : ''
      }
      
      nodata = {
        //display if there is no favourite
        (favouriteStatus === 'SUCCESS' && favouriteMovies.results.length < 1)
        ? <NoData/>
        : null
      }

      pagination = {
        //pagination
        favouriteStatus === 'SUCCESS' ? favouriteMovies.total_results > 20 ? <Pagination 
                                      searchMore={(page) => getAllFavourites(page)}
                                      page={favouriteMovies.page} 
                                      pages={favouriteMovies.total_pages}
                                      /> 
                : ''
                : ''
      }

    />
  );
}

const mapStateToProps = (state) => (
  {
    favouriteStatus: state.status.favourite,
    favouriteMovies: state.favourites.favouriteMovies,
    watchlistActionStatus: state.status.watchlistAction,
    favouriteActionStatus: state.status.favouriteAction
  }
);
  
const mapDispatchToProps = (dispatch) => (
  {
    getAllFavourites: (page) => dispatch(getAllFavourites(page)),
    removeFavourite: (payload) => dispatch(removeFromFavourite(payload)),
    addToWatchList: (payload) => dispatch(addToWatchList(payload)),
    removeFromWatchList: (payload) => dispatch(removeFromWatchList(payload)),
    resetStatus: () => dispatch(resetStatus())
  }
)

export default connect(mapStateToProps, mapDispatchToProps) (Favourites);
