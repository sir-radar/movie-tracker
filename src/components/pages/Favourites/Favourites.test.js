import React from 'react';
import { MemoryRouter } from "react-router-dom";
import { render, cleanup } from '@testing-library/react';
import Favourites from './Favourites';
import "@testing-library/jest-dom/extend-expect";
import renderer from "react-test-renderer";
import { Provider } from 'react-redux';
import thunk from "redux-thunk";
import configureStore from "redux-mock-store";


const buildStore = configureStore([thunk]);

//reducers
import auth from '../../../store/reducers/authReducer';



afterEach(cleanup)

const renderComponent = (states) => {
  //mock store
  const store = buildStore({...states});
  const div = document.createElement("div");
  return render(<Provider store={store}><MemoryRouter><Favourites /></MemoryRouter></Provider>, div)
}

const renderComponentForSnapshot = (states) => {
  const store = buildStore({...states});
  return <Provider store={store}><MemoryRouter><Favourites /></MemoryRouter></Provider>
}

it("renders without crashing", async () => {

  const states = {
    auth,
    watchlists:{
      watchListsID:[1,2,3],
    },
    status:{
      favourite: 'PENDING',
      watchlistAction:'',
      favouriteAction:''
    },
    favourites:{
      favouriteMovies:{},
      favouriteMoviesID:[]
    }
  }

  renderComponent(states)
  
})

it("renders nav component", () => {

  const states = {
    auth,
    watchlists:{
      watchListsID:[1,2,3],
    },
    status:{
      favourite: 'PENDING',
      watchlistAction:'',
      favouriteAction:''
    },
    favourites:{
      favouriteMovies:{},
      favouriteMoviesID:[]
    }
  }

  const {queryByTestId} = renderComponent(states)
  expect(queryByTestId("nav")).not.toBeNull()
  expect(queryByTestId("movie-card")).toBeNull()
  expect(queryByTestId("pagination")).toBeNull()
})

it("renders loading component on pending request", () => {

  const states = {
    auth,
    watchlists:{
      watchListsID:[1,2,3],
    },
    status:{
      favourite: 'PENDING',
      watchlistAction:'',
      favouriteAction:''
    },
    favourites:{
      favouriteMovies:{},
      favouriteMoviesID:[]
    }
  }

  const {queryByTestId} = renderComponent(states)
  expect(queryByTestId("nav")).not.toBeNull()
  expect(queryByTestId("loader")).not.toBeNull()
  expect(queryByTestId("movie-card")).toBeNull()
  expect(queryByTestId("pagination")).toBeNull()

})

it("renders Movie Card component on successful request with at least one result", () => {

  const states = {
    auth,
    watchlists:{
      watchListsID:[1,2,3],
    },
    status:{
      favourite: 'SUCCESS',
      watchlistAction:'',
      favouriteAction:''
    },
    favourites:{
      favouriteMovies:{
        results:[
          {id: 1, title:'hey', overview:'hey'}
        ]
      },
      favouriteMoviesID:[]
    }
  }

  
  const {queryByTestId} = renderComponent(states)
  expect(queryByTestId("loader")).toBeNull()
  expect(queryByTestId("error")).toBeNull()
  expect(queryByTestId("error-message")).toBeNull()
  expect(queryByTestId("nodata-component")).toBeNull()
  expect(queryByTestId("movie-card")).not.toBeNull()
  expect(queryByTestId("pagination")).toBeNull()

})


it("renders NoData component on successful request with empty data", () => {
  const states = {
    auth,
    watchlists:{
      watchListsID:[1,2,3],
    },
    status:{
      favourite: 'SUCCESS',
      watchlistAction:'',
      favouriteAction:''
    },
    favourites:{
      favouriteMovies:{
        results:[]
      },
      favouriteMoviesID:[]
    }
  }

  
  const {queryByTestId} = renderComponent(states)
  expect(queryByTestId("loader")).toBeNull()
  expect(queryByTestId("error")).toBeNull()
  expect(queryByTestId("error-message")).toBeNull()
  expect(queryByTestId("nodata-component")).not.toBeNull()
  expect(queryByTestId("movie-card")).toBeNull()
  expect(queryByTestId("pagination")).toBeNull()

})

it("renders Error component on failed request", () => {
  
  const states = {
    auth,
    watchlists:{
      watchListsID:[1,2,3],
    },
    status:{
      favourite: 'ERROR',
      watchlistAction:'',
      favouriteAction:''
    },
    favourites:{
      favouriteMovies:{
        results:[]
      },
      favouriteMoviesID:[]
    }
  }

  
  const {queryByTestId} = renderComponent(states)
  expect(queryByTestId("loader")).toBeNull()
  expect(queryByTestId("error")).not.toBeNull()
  expect(queryByTestId("error-message")).toBeNull()
  expect(queryByTestId("nodata-component")).toBeNull()
  expect(queryByTestId("movie-card")).toBeNull()
  expect(queryByTestId("pagination")).toBeNull()

})

it("renders Pagination component on successful request with more than 20 results", () => {
  const states = {
    auth,
    watchlists:{
      watchListsID:[1,2,3],
    },
    status:{
      favourite: 'SUCCESS',
      watchlistAction:'',
      favouriteAction:''
    },
    favourites:{
      favouriteMovies:{
        results:[
          {id: 1, title:'hey', overview:'hey'}
        ],
        total_results: 21,
        page: 1,
        total_pages: 2
      },
      favouriteMoviesID:[]
    }
  }

  
  const {queryByTestId, queryAllByTestId} = renderComponent(states)
  expect(queryByTestId("loader")).toBeNull()
  expect(queryByTestId("error")).toBeNull()
  expect(queryByTestId("error-message")).toBeNull()
  expect(queryByTestId("nodata-component")).toBeNull()
  expect(queryAllByTestId("movie-card").length).toBe(states.favourites.favouriteMovies.results.length)
  expect(queryByTestId("pagination")).not.toBeNull()

})

it("renders Error Message component on failed add/remove favourite or watchlist request", () => {
  const states = {
    auth,
    watchlists:{
      watchListsID:[1,2,3],
    },
    status:{
      favourite: 'SUCCESS',
      watchlistAction:'ERROR',
      favouriteAction:'ERROR'
    },
    favourites:{
      favouriteMovies:{
        results:[
          {id: 1, title:'hey', overview:'hey'}
        ],
        total_results: 21,
        page: 1,
        total_pages: 2
      },
      favouriteMoviesID:[]
    }
  }

  
  const {queryByTestId, queryAllByTestId} = renderComponent(states)
  expect(queryByTestId("loader")).toBeNull()
  expect(queryByTestId("error")).toBeNull()
  expect(queryByTestId("error-message")).not.toBeNull()
  expect(queryByTestId("nodata-component")).toBeNull()
  expect(queryAllByTestId("movie-card").length).toBe(states.favourites.favouriteMovies.results.length)
  expect(queryByTestId("pagination")).not.toBeNull()

})


it("matches snapshot", () => {
  const states = {
    auth,
    watchlists:{
      watchListsID:[1,2,3],
    },
    status:{
      favourite: 'PENDING',
      watchlistAction:'',
      favouriteAction:''
    },
    favourites:{
      favouriteMovies:{},
      favouriteMoviesID:[]
    }
  }
  const tree = renderer.create(renderComponentForSnapshot(states)).toJSON();
  expect(tree).toMatchSnapshot();
})