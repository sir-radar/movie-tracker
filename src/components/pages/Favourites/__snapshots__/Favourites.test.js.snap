// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Favourite Page Component matches snapshot 1`] = `
Array [
  <nav
    className="d-flex justify-content-center"
    data-testid="nav"
  >
    <a
      className="p-3 primary-text-color active"
      data-testid="search"
      href="/"
      onClick={[Function]}
    >
      Search
    </a>
    <a
      className="p-3 primary-text-color"
      data-testid="favourite"
      href="/favourites"
      onClick={[Function]}
    >
      Favorites
    </a>
    <a
      className="p-3 primary-text-color"
      data-testid="watchlater"
      href="/watchlist"
      onClick={[Function]}
    >
      WatchLater
    </a>
  </nav>,
  <main
    className="mt-4 p-3 col-10 col-md-10 offset-1 offset-md-1"
  >
    <div
      className="row"
    >
      <section
        className="loading mt-5 mx-auto"
        data-testid="loader"
      >
        <i
          className="fa fa-spinner fa-5x fa-spin"
        />
      </section>
      
      
    </div>
  </main>,
]
`;
